apollo-router/src/plugins/rhai.rs

```diff
@@ -164,386 +270,545 @@ impl Plugin for Rhai {
+                x.operation_name
+                    .clone()
+                    .map_or(Dynamic::from(()), Dynamic::from)
+            })
+            .register_set("operation_name", |x: &mut Request, value: &str| {
+                x.operation_name = Some(value.to_string());
+            })
+            // Request.variables
+            .register_get_result("variables", |x: &mut Request| {
+                to_dynamic(x.variables.clone())
+            })
+            /* XXX CANNOT DO BECAUSE variables is Arc
+            .register_set_result("variables", |x: &mut Request, om: Map| {
+                x.variables = from_dynamic(om)
+            })
+            */
+            // Request.extensions
+            .register_get_result("extensions", |x: &mut Request| {
+                to_dynamic(x.extensions.clone())
+            })
+            .register_set_result("extensions", |x: &mut Request, om: Map| {
+                let v = rhai::format_map_as_json(&om);
+                x.extensions = serde_json::from_str(&v).map_err(|e| e.to_string())?;
+                Ok(())
+            })
+            // Register a series of logging functions
+            .register_fn("log_trace", |x: &str| {
+                tracing::trace!("{}", x);
+            })
+            .register_fn("log_debug", |x: &str| {
+                tracing::debug!("{}", x);
+            })
+            .register_fn("log_info", |x: &str| {
+                tracing::info!("{}", x);
+            })
+            .register_fn("log_warn", |x: &str| {
+                tracing::warn!("{}", x);
+            })
+            .register_fn("log_error", |x: &str| {
+                tracing::error!("{}", x);
+            })
+            // Register a function for printing to stderr
+            .register_fn("eprint", |x: &str| {
+                eprintln!("{}", x);
+            })
+            // Default representation in rhai is the "type", so
+            // we need to register a to_string function for all our registered
+            // types so we can interact meaningfully with them.
+            .register_fn("to_string", |x: &mut Option<HeaderName>| -> String {
+                match x {
+                    Some(v) => v.to_string(),
+                    None => "None".to_string(),
+                }
+            })
+            .register_fn("to_string", |x: &mut HeaderName| -> String {
+                x.to_string()
+            })
+            .register_fn("to_string", |x: &mut HeaderValue| -> String {
+                x.to_str().map_or("".to_string(), |v| v.to_string())
+            })
+            .register_fn("to_string", |x: &mut HeaderMap| -> String {
+                let mut msg = String::new();
+                for pair in x.iter() {
+                    let line = format!(
+                        "{}: {}",
+                        pair.0,
+                        pair.1.to_str().map_or("".to_string(), |v| v.to_string())
```

Comment:
```suggestion
                        pair.1.to_str().map_or("<invalid_header_value>".to_string(), |v| v.to_string())
```

Reply:
This is one of those where there is no good single answer. The interface requires a string and you can't raise an error. I can see that specifying something like you suggest might be better than an empty string... If someone is passing around header values that can't be converted to string representation, there are probably bigger issues. :)

I'm going to leave as is for now, but maybe we can re-visit this post GA and see if a clear, good solution presents itself?

Reply:
SGTM